src/common/lib/util/defaults.ts

```diff
@@ -119,15 +121,34 @@ export function getHttpScheme(options: ClientOptions): string {
   return options.tls ? 'https://' : 'http://';
 }
 
-// construct environment fallback hosts as per RSC15i
-export function environmentFallbackHosts(environment: string): string[] {
-  return [
-    environment + '-a-fallback.ably-realtime.com',
-    environment + '-b-fallback.ably-realtime.com',
-    environment + '-c-fallback.ably-realtime.com',
-    environment + '-d-fallback.ably-realtime.com',
-    environment + '-e-fallback.ably-realtime.com',
-  ];
+export function getEndpointHostname(endpoint: string): string {
+  if (endpoint.includes('.') || endpoint.includes('::') || endpoint.includes('localhost')) {
+    return endpoint;
+  }
+
+  if (endpoint.startsWith('nonprod:')) {
+    const root = endpoint.replace('nonprod:', '');
+    return `${root}.realtime.ably-nonprod.net`;
+  }
+
+  return `${endpoint}.realtime.ably.net`;
+}
+
+export function getEndpointFallbackHosts(endpoint: string): string[] {
```

Comment:
Please can you annotate with the relevant spec points?

Reply:
Done